### PR TITLE
Add TimeSync.getStnGraph()

### PR DESCRIPTION
**Pull request by jamesmontalvo3**
This PR adds the ability to get the STN graph data. The GIF below shows that data visualized with D3.js, but that visualization is not part of the PR.

time-sync-graph mp4-20200116152030

To inspect the output of `TimeSync.getStnGraph()`, do the following in the browser console:

```javascript
maestro.react.app.state.procedure.timeSync.getStnGraph();
```

To create the D3.js force-directed-graph as shown in the GIF above, punch this into the console:

```javascript
function loadScript(src) {
    var ele = document.createElement("script");
    ele.setAttribute("src",src);
    document.head.appendChild(ele)
}
document.getElementById('root').innerHTML = '<svg width="1800" height="900"></svg>';
loadScript('https://d3js.org/d3.v4.min.js');
setTimeout(function(){
    loadScript('https://d3js.org/d3-selection-multi.v1.js');
}, 1000)

function makeTheGraph() {
    var colors = d3.scaleOrdinal(d3.schemeCategory10);

    var svg = d3.select("svg"),
        width = +svg.attr("width"),
        height = +svg.attr("height"),
        node,
        link;

    svg.append('defs').append('marker')
        .attrs({'id':'arrowhead',
            'viewBox':'-0 -5 10 10',
            'refX':13,
            'refY':0,
            'orient':'auto',
            'markerWidth':13,
            'markerHeight':13,
            'xoverflow':'visible'})
        .append('svg:path')
        .attr('d', 'M 0,-5 L 10 ,0 L 0,5')
        .attr('fill', '#999')
        .style('stroke','none');

    var simulation = d3.forceSimulation()
        .force("link", d3.forceLink().id(function (d) {return d.id;}).distance(400).strength(1))
        .force("charge", d3.forceManyBody())
        .force("center", d3.forceCenter(width / 2, height / 2));

    function update(links, nodes) {
        link = svg.selectAll(".link")
            .data(links)
            .enter()
            .append("line")
            .attr("class", "link")
            .attr('marker-end','url(#arrowhead)')

        link.append("title")
            .text(function (d) {return d.action;});

        edgepaths = svg.selectAll(".edgepath")
            .data(links)
            .enter()
            .append('path')
            .attrs({
                'class': 'edgepath',
                'fill-opacity': 0,
                'stroke-opacity': 0,
                'id': function (d, i) {return 'edgepath' + i}
            })
            .style("pointer-events", "none");

        edgelabels = svg.selectAll(".edgelabel")
            .data(links)
            .enter()
            .append('text')
            .style("pointer-events", "none")
            .attrs({
                'class': 'edgelabel',
                'id': function (d, i) {return 'edgelabel' + i},
                'font-size': 10,
                'fill': '#aaa'
            });

        edgelabels.append('textPath')
            .attr('xlink:href', function (d, i) {return '#edgepath' + i})
            .style("text-anchor", "middle")
            .style("pointer-events", "none")
            .attr("startOffset", "50%")
            .text(function (d) {return d.action});

        node = svg.selectAll(".node")
            .data(nodes)
            .enter()
            .append("g")
            .attr("class", "node")
            .call(d3.drag()
                    .on("start", dragstarted)
                    .on("drag", dragged)
                    //.on("end", dragended)
            );

        node.append("circle")
            .attr("r", 5)
            .style("fill", function (d, i) {return colors(i);})

        node.append("title")
            .text(function (d) {return d.id;});

        node.append("text")
            .attr("dy", -3)
            .text(function (d) {return d.label;});

        simulation
            .nodes(nodes)
            .on("tick", ticked);

        simulation.force("link")
            .links(links);
    }

    function ticked() {
        link
            .attr("x1", function (d) {return d.source.x;})
            .attr("y1", function (d) {return d.source.y;})
            .attr("x2", function (d) {return d.target.x;})
            .attr("y2", function (d) {return d.target.y;});

        node
            .attr("transform", function (d) {return "translate(" + d.x + ", " + d.y + ")";});

        edgepaths.attr('d', function (d) {
            return 'M ' + d.source.x + ' ' + d.source.y + ' L ' + d.target.x + ' ' + d.target.y;
        });

        edgelabels.attr('transform', function (d) {
            if (d.target.x < d.source.x) {
                var bbox = this.getBBox();

                rx = bbox.x + bbox.width / 2;
                ry = bbox.y + bbox.height / 2;
                return 'rotate(180 ' + rx + ' ' + ry + ')';
            }
            else {
                return 'rotate(0)';
            }
        });
    }

    function dragstarted(d) {
        if (!d3.event.active) simulation.alphaTarget(0.3).restart()
        d.fx = d.x;
        d.fy = d.y;
    }

    function dragged(d) {
        d.fx = d3.event.x;
        d.fy = d3.event.y;
    }

    //    function dragended(d) {
    //        if (!d3.event.active) simulation.alphaTarget(0);
    //        d.fx = undefined;
    //        d.fy = undefined;
    //    }

    const graph = maestro.react.app.state.procedure.timeSync.getStnGraph();
    update(graph.edges, graph.nodes);
    window.graph = graph;
    // add this CSS: .link { stroke: #999; stroke-opacity: .6; stroke-width: 1px; }
}

console.log('okay about to make a graph...wait while stuff loads');
setTimeout(makeTheGraph, 2000);
```

### CODE REVIEW

File: app/model/TimeSync.js [hunk 1]
```diff
@@ -427,4 +427,135 @@ module.exports = class TimeSync {
 		return presentActors;
 	}
 
+	/**
+	 * @return {Object}  Object with nodes and edges, in form:
```
Comment (cameronwp) [inline app/model/TimeSync.js]: Nit: as much as I love the detail here, I'd argue that you could truncate the example. Maybe shorten `nodes` and `links` to just two or three items?

File: app/model/TimeSync.js [hunk 2]
```diff
@@ -427,4 +427,135 @@ module.exports = class TimeSync {
...
+	 *                       {
+	 *                         source: 6, target: 7, minutes: 0,
+	 *                         actor: 'EV1 --> EV2 sync offset for procedure end'
+	 *                       },
+ 	 *                     ]
+	 *                   };
+	 */
+	getStnGraph() {
+
+		// each task will have a node for _each actor on the task_
```
Comment (cameronwp) [inline app/model/TimeSync.js]: I like the use of markdown in code comments :)

File: app/model/TimeSync.js [hunk 3]
```diff
@@ -427,4 +427,135 @@ module.exports = class TimeSync {
...
+		const nodes = [];
+		const edges = [];
+
+		const actorEndpoints = {};
+
+		const taskToIndexMap = new Map(); // for a given task get its 'nodes' indices for each actor
+		this.tasks.forEach((task) => {
+			const actorToNode = {};
+			for (const actor in task.actorRolesDict) {
+				const nodeId = nodes.length;
```
Comment (cameronwp) [inline app/model/TimeSync.js]: clever!

**Comment by cameronwp** [comment]
@jamesmontalvo3 I'm trusting your judgement wrt using models to generate the graph. You know them much better than I do. The strategy here generally looks sound. A couple tests and I'd call this one good.

**Comment by cameronwp** [comment]
Upon further testing, I've got a request. We need a guarantee that each node id is unique. If two actors perform the same task, each needs their own node with its own id.

File: app/model/TimeSync.js [hunk 4]
```diff
@@ -427,4 +427,135 @@ module.exports = class TimeSync {
...
+ 	 *                       { id: 3, label: 'Start of Some Task for EV2' },
+ 	 *                       { id: 4, label: 'Start of Cleanup/Ingress for EV1' },
+ 	 *                       { id: 5, label: 'Start of Cleanup/Ingress for EV2' },
+ 	 *                       { id: 6, label: 'End of procedure for EV1' },
+ 	 *                       { id: 7, label: 'End of procedure for EV2' }
+ 	 *                     ],
+ 	 *                     links: [
+ 	 *                       { source: 0, target: 2, minutes: 60, actor: 'EV1' },
+	 *                       {
+	 *                         source: 0, target: 1, minutes: 0,
```
Comment (cameronwp) [inline app/model/TimeSync.js]: on second look, it looks like you changed the `value` interval to `minutes`. what's easier for you? we just need to be explicit about which format is easier for you. I can handle either.

Comment (cameronwp) [reply app/model/TimeSync.js]: Though the current implementation with `minutes` looks like there aren't really upper or lower intervals between tasks?

**Comment by jamesmontalvo3** [comment]
> A couple tests and I'd call this one good.

The graph this generates is pretty complicated. I think the best way to "test" it is for you to validate the output of a couple graphs, make sure it's what you want, then just copy that data structure into a test and make sure it keeps matching it in the future (it will by-definition match it on day zero).

> Upon further testing, I've got a request. We need a guarantee that each node id is unique. If two actors perform the same task, each needs their own node with its own id.

Is it not doing this currently?

**Comment by cameronwp** [comment]
I'm working on testing the integration with my STNs. Let's merge this and I'll submit follow up PRs if any modifications are necessary.

Yes, in the current implementation node IDs are unique. I just want to be explicit that this is a requirement from my STN implementation and cannot change in the future.

**Pull request merged**